### DISCUSSION TITLE
(not a security issue in pretix) kdevtmpfsi and kinsing malware

### DISCUSSION THREAD

**Discussion by jxs714** [pretix/pretix#3670] [Q&A] [answered]
Hey guys,

Just a heads up. The official docker for pretix comes with a default root username and password. This is a serious security issue because outside users/bots can scan the docker API to get into your docker as root. If the user and password is root, that is instant access.

**Testing I've done**
Issue tested on
- Postgresql 13 and 16 versions
- Latest Pretix 2023.8.0

I noticed upon testing that my Postgresql (both 13 and 16 versions) would get infected with a malware and process called kdevtmpfsi. This is a crypto currency malware that uses your server as a crypto mining machine. This malware was using 50%+ of my cpu usage. I've ran tests for hours assuming that the reason had to be that they somehow logged into my postgresql root account. I changed my root password many times and this was not the case. Every time you kill the process, it starts up again minutes later or after you restart postgresql service.

**Steps for solution**
If infected first do the following:
• Go into command line and check if the malware is in other places of your server (besides the /tmp folder) by using the "Find" feature. You can find instructions on how to do this here: [https://www.freecodecamp.org/news/how-to-search-files-in-the-linux-terminal/#:~:text=By%20passing%20the%20name%20of,the%20location%20of%20the%20file.&text=But%20remember%20the%20%2Dname%20flag,use%20the%20%2Diname%20flag%20instead.] For me, it only infected the /tmp folder with a file named kdevtmpfsi. Other users with this same docker vulnerability have had different experiences with weird file names in the /tmp folder.
• Run the following commands to remove the malware files from server

sudo rm -f /tmp/kdevtmpfsi
sudo rm -f /var/tmp/kinsing
sudo rm -f /tmp/kinsing

KILL the services
pkill kdevtmpfsi
pkill kinsing

• Stop postgresql service. (to prevent the malware from returning)
• Go into the pretix docker container with bash (Example: docker exec -it --user root NAMEOFDOCKERCONTAINER bash
• Change your docker container root password with the command: Passwd
• Return to the main server command prompt and leave the docker bash command with the command "exit"
• Start postgresql service
• Next, you also want to be sure to modify your admin postgresql password with a strong password for increased security.

Once this is all done, you should be all set to go.

**Comment by raphaelm** [comment]
> This is a serious security issue because outside users/bots can scan the docker API to get into your docker as root.

How would an outside user get access to your docker API in the first place?

**Reply by jxs714** [reply]
Hello,

Brute force credentials. In this case, default user and password for docker containers are usually username "root" and password "root". That is an easy way for a bot to get in and infect with crypto malware. They can start by scanning the system for docker through docker api & possible open ports. Its a very easy way to get into your docker root if your login is the default docker login. Because I didn't install postgresql in another container, and instead on my host system, it somehow targeted and infected my postgresql server in my host system and created a file in the /tmp folder. I know for a fact that the issue was the docker root login because once I changed the password in bash, hours have passed and the malware has not made its way back in. I will continue to do more testing and checking. If the malware returns then my research and testing was incorrect. If so, will update this forum

**Reply by raphaelm** [reply]
> Brute force credentials. In this case, default user and password for docker containers are usually username "root" and password "root".

But how would you access that? From the network? Why would it be accessible through the network?

> They can start by scanning the system for docker through docker api & possible open ports

Why would the docker api be exposed to an external attacker?

**Reply by jxs714** [reply]
Hi Raphael,

Docker has been known to require extra steps to secure. Typically bots will aim for the api and find what ports are open. You would have to find more research on it via google..

However, I believe I may have found the actual culprit and issue. After thorough testing for the past 24-48 hours, the issue where malware infected the host systems /tmp folder with the malware came up right after installing set of plugins. The plugin that was the only 3rd party plugin was Batch emailer here: https://marketplace.pretix.eu/products/batch_emailer/ - I've renamed this ticket title so that it doesn't damage the integrity of the official docker. I didn't think about this when I started the ticket, my apologies for this.

I will continue to monitor to see if the malware appears only using the official plugins. Will update this forum on my findings and results for the next 1-2 days.

Thank you!

**Reply by raphaelm** [reply, marked as answer]
Hi,

please don't take this personally, but I want to make this clear in case other people are reading this discussion and worrying about a security issue in pretix: From your replies, it seems that you have no idea what you are talking about. Please hire someone who knows how to securely administer a linux system to avoid putting your users at risk.

> Docker has been known to require extra steps to secure. Typically bots will aim for the api and find what ports are open. 

This is wrong. The docker API, by default, does not listen on a port that is reachable from outside. By default, it only listens on a socket at ``/var/run/docker.sock`` that can be accessed only be the ``docker`` unix group. So an attacker from the internet cannot access the docker API, unless you actively opened it up.

> Its a very easy way to get into your docker root if your login is the default docker login.

Only if you screwed up your docker configuration and your firewall configuration.

> The official docker for pretix comes with a default root username and password.

This is wrong. The root user in the docker container has no password set.

> If the user and password is root, that is instant access.

This is wrong. If a user *has* access to the docker API (by default: local user in the docker group, see above), they can access any container no matter what the password is. If they do not have access to the docker API, they don't have access regardless of the password, since the container exposes no ssh or other login service.

>  I've ran tests for hours assuming that the reason had to be that they somehow logged into my postgresql root account.

That seems more likely and is a common path of entry for the kdevtmpfsi malware. It also means your postgresql was probably set up insecurely, why would it be exposed on the public network? PostgreSQL does not accept authenticated connections from random sources with default configuration.

> I changed my root password many times and this was not the case. Every time you kill the process, it starts up again minutes later or after you restart postgresql service.

That might be because the malware you cited is known to add itself as a cronjob.

> However, I believe I may have found the actual culprit and issue. After thorough testing for the past 24-48 hours, the issue where malware infected the host systems /tmp folder with the malware came up right after installing set of plugins.

This seems like a complete coincidence.

> he plugin that was the only 3rd party plugin was Batch emailer here: [marketplace.pretix.eu/products/batch_emailer](https://marketplace.pretix.eu/products/batch_emailer/)

I can't be 100% certain, but I've now read through the plugin code and I don't think it includes malware.

> I will continue to monitor to see if the malware appears only using the official plugins. Will update this forum on my findings and results for the next 1-2 days.

Yes, you will see it again, regardless of plugins, if it infected your cronjobs.

I'd *strongly* advise you to

- set up an entirely new server from scratch
- make sure no ports are available from outside that shouldn't be (do a portscan)
- find someone competent to double-check it

**Reply by jxs714** [reply]
Hi 

Thanks for the response. My apologies for the initial post, did not intend any harm at all, just trying to help you guys and the community just in case there was anything wrong somewhere outside of you guys. To reference, if anyone has this issue, be sure to do the following:

- Change postgresql password to a strong password
- Lock postgresql from letting outside users access your database remotely once configured
- If infected, make sure to check both CRONTAB in linux and the crontab file location of postgresql at /var/spool/cron/crontabs 

For some reason they hid the cron from crontab and scheduled a malicious script to download and run in cron format in the postgresql file in that folder, which is why I couldn't see them re-infecting.

Thanks and keep up the great work Raphael + Team

**Reply by c0de-bender** [reply]
@jxs714 I'm using Pretix too. When I read the question title, I was worried or let's say "curious" in a worrying way. So I came here and read the whole discussion. I'm happy to hear that you could fix the problem and that it's not a problem with Pretix (or one of its plugins). :)

I have a wish: Could you change the title of the question again? Like to something that makes it clear that it's not a Pretix issue? This way other people don't need to read through the whole discussion to find out there's no problem with Pretix. :wink:

**Reply by raphaelm** [reply]
I updated the discussion title
